lib/seadsa/ShadowMem.cc

```diff
@@ -1177,6 +1189,59 @@ void ShadowMemImpl::visitMemhavoc(CallSite &CS) {
   associateConcretePtr(memUse, ptr, &callInst);
 }
 
+void ShadowMemImpl::visitIsModified(CallSite &CS) {
+  LOG("dsa.ismodified", errs() << "Visiting isModified \n");
+  auto &callInst = *CS.getInstruction();
+  /* isModified definition:
+     bool @sea_is_modified(i8*)
+     operand 0 is the pointer to be marked
+  */
+  auto &ptr = *CS.getArgOperand(0);
+  if (!m_graph->hasCell(ptr)) {
+    LOG("dsa.ismodified", errs() << "no cell for: " << ptr << "\n");
+    return;
+  }
+  const dsa::Cell &cell = m_graph->getCell(ptr);
+  if (cell.isNull()) {
+    LOG("dsa.ismodified", errs() << "cell is null for: " << ptr << "\n");
+    return;
+  }
+
+  m_B->SetInsertPoint(&callInst);
+  CallInst &memUse =
+      mkShadowLoad(*m_B, cell, dsa::getTypeSizeInBytes(*ptr.getType(), *m_dl));
+  associateConcretePtr(memUse, ptr, &callInst);
+}
+
+void ShadowMemImpl::visitResetModified(CallSite &CS) {
+  LOG("dsa.resetmodified", errs() << "Visiting resetModified \n");
+  auto &callInst = *CS.getInstruction();
+  /* resetModified definition:
+     bool @sea.reset_modified(i8*)
+     operand 0 is the pointer to be marked
+  */
+  auto &ptr = *CS.getArgOperand(0);
+  if (!m_graph->hasCell(ptr)) {
+    LOG("dsa.resetmodified", errs() << "no cell for: " << ptr << "\n");
+    return;
+  }
+  const dsa::Cell &cell = m_graph->getCell(ptr);
+  if (cell.isNull()) {
+    LOG("dsa.resetmodified", errs() << "cell is null for: " << ptr << "\n");
+    return;
+  }
+
+  if (cell.getOffset() == 0) {
+    m_B->SetInsertPoint(&callInst);
+    CallInst &memDef = mkShadowStore(
+        *m_B, cell, dsa::AllocSiteInfo::getAllocSiteSize(callInst));
+    associateConcretePtr(memDef, callInst, &callInst);
+  } else {
+    errs() << "WARNING: skipping sea.reset_modified instrumentation since "
```

Comment:
`offset == 0` means that pointer is to the beginning of allocated object. I don't think we restrict flag reset to beginning of an object. Even if we do, `offset == 0` implies that pointer is to an object, but not the other way around.

Reply:
Ack.